src/gui/qgscredentialdialog.cpp

```diff
@@ -45,7 +56,58 @@ QgsCredentialDialog::QgsCredentialDialog( QWidget *parent, Qt::WindowFlags fl )
   connect( this, &QgsCredentialDialog::credentialsRequestedMasterPassword,
            this, &QgsCredentialDialog::requestCredentialsMasterPassword,
            Qt::BlockingQueuedConnection );
-  mOkButton = buttonBox->button( QDialogButtonBox::Ok );
+
+  // Setup ignore button
+  mIgnoreButton->setToolTip( tr( "All requests for this connection will be automatically rejected" ) );
+  QMenu *menu = new QMenu( mIgnoreButton );
+  QAction *ignoreTemporarily = new QAction( tr( "Ignore for 10 seconds" ), menu );
+  ignoreTemporarily->setToolTip( tr( "All requests for this connection will be automatically rejected for 10 seconds" ) );
+  QAction *ignoreForSession = new QAction( tr( "Ignore for session" ), menu );
+  ignoreForSession->setToolTip( tr( "All requests for this connection will be automatically rejected for the duration of the current session" ) );
+  menu->addAction( ignoreTemporarily );
+  menu->addAction( ignoreForSession );
+  connect( ignoreTemporarily, &QAction::triggered, [ = ]
+  {
+    mIgnoreMode = IgnoreTemporarily;
+    //mIgnoreButton->setText( ignoreTemporarily->text() );
+    mIgnoreButton->setToolTip( ignoreTemporarily->toolTip() );
+  } );
+  connect( ignoreForSession, &QAction::triggered, [ = ]
+  {
+    mIgnoreMode = IgnoreForSession;
+    //mIgnoreButton->setText( ignoreForSession->text() );
+    mIgnoreButton->setToolTip( ignoreForSession->toolTip() );
+  } );
+  mIgnoreButton->setText( mIgnoreMode == IgnoreTemporarily ? ignoreTemporarily->text() : ignoreForSession->text() );
+  mIgnoreButton->setToolTip( mIgnoreMode == IgnoreTemporarily ? ignoreTemporarily->toolTip() : ignoreForSession->toolTip() );
+  mIgnoreButton->setMenu( menu );
+  mIgnoreButton->setMaximumHeight( mOkButton->sizeHint().height() );
+
+  // Connect ok and cancel buttons
+  connect( mOkButton, &QPushButton::clicked, this, &QgsCredentialDialog::accept );
+  connect( mCancelButton, &QPushButton::clicked, this, &QgsCredentialDialog::reject );
+
+  // Keep a cache of ignored connections, and ignore them for 10 seconds.
+  connect( mIgnoreButton, &QPushButton::clicked, [ = ]( bool )
+  {
+    const QString realm { labelRealm->text() };
+    {
+      QMutexLocker locker( &sIgnoredConnectionsCacheMutex );
+      // Insert the realm in the cache of ignored connections
+      sIgnoredConnectionsCache->insert( realm );
+    }
+    if ( mIgnoreMode == IgnoreTemporarily )
```

Comment:
Use a switch here instead of if

Reply:
why? I'm not handling the other case.

Reply:
Makes it much easier to extend this in future, because you get compiler warnings wherever a new enum value is added. It's one of the core C++ guidelines - you should always use switches for enum based logic.

Reply:
I fully agree in the general case where a switch is better than a chin of if/else if, but here it would just make the code much less readable because the enum is just a replacement for a bool.